src/frontend/src/optimizer/rule/pull_up_correlated_predicate.rs

```diff
@@ -0,0 +1,136 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::{Either, Itertools};
+
+use super::super::plan_node::*;
+use super::{BoxedRule, Rule};
+use crate::expr::{CorrelatedInputRef, Expr, ExprImpl, ExprRewriter, InputRef};
+use crate::optimizer::PlanRef;
+use crate::utils::Condition;
+
+/// This rule is for pattern: Apply->Project->Filter.
+///
+/// To unnest, we just pull predicates contain correlated variables in Filter into Apply, and
+/// convert it into corresponding type of Join.
+pub struct PullUpCorrelatedPredicate {}
+impl Rule for PullUpCorrelatedPredicate {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let apply = plan.as_logical_apply()?;
+
+        let right = apply.right();
+        let project = right.as_logical_project()?;
+        let (mut exprs, mut expr_alias, _) = project.clone().decompose();
+        let begin = exprs.len();
+
+        let input = project.input();
+        let filter = input.as_logical_filter()?;
+
+        // Split predicates in LogicalFilter into correlated expressions and uncorrelated
+        // expressions.
+        let (cor_exprs, uncor_exprs) =
+            filter
+                .predicate()
+                .clone()
+                .into_iter()
+                .partition_map(|expr| {
+                    let mut rewriter = Rewriter {
+                        has_correlated_input_ref: false,
+                        input_refs: vec![],
+                        index: exprs.len() + apply.left().schema().fields().len(),
+                    };
+                    let rewritten_expr = rewriter.rewrite_expr(expr.clone());
+                    if rewriter.has_correlated_input_ref {
+                        // Append input_refs in expression which contains correlated_input_ref to
+                        // `exprs` used to construct new LogicalProject.
+                        exprs.extend(
+                            rewriter
+                                .input_refs
+                                .drain(..)
+                                .map(|input_ref| input_ref.into()),
+                        );
+                        Either::Left(rewritten_expr)
+                    } else {
+                        Either::Right(expr)
+                    }
+                });
+
+        // TODO: remove LogicalFilter with always true condition.
+        let filter = LogicalFilter::new(
+            filter.input(),
+            Condition {
+                conjunctions: uncor_exprs,
+            },
+        );
+
+        // Add columns involved in expressions removed from LogicalFilter to LogicalProject.
+        let end = exprs.len();
+        expr_alias.extend(vec![None; end - begin].into_iter());
```

Comment:
The comment is misleading because they are added above, not here. Why not extending `expr` and `expr_alias` at the same time?